packages/contracts-bedrock/src/cannon/MIPS2.sol

```diff
@@ -0,0 +1,642 @@
+
+            // Handle syscall separately
+            // syscall (can read and write)
+            if (opcode == 0 && fun == 0xC) {
+                return handleSyscall(_localContext);
+            }
+
+            // Exec the rest of the step logic
+            st.CpuScalars memory cpu = getCpuScalars(thread);
+            (state.memRoot) = ins.execMipsCoreStepLogic({
+                _cpu: cpu,
+                _registers: thread.registers,
+                _memRoot: state.memRoot,
+                _memProofOffset: MIPSMemory.memoryProofOffset(MEM_PROOF_OFFSET, 1),
+                _insn: insn,
+                _opcode: opcode,
+                _fun: fun
+            });
+            setStateCpuScalars(thread, cpu);
+            updateCurrentThreadRoot();
+            return outputState();
+        }
+    }
+
+    function handleSyscall(bytes32 _localContext) internal returns (bytes32 out_) {
+        unchecked {
+            // Load state from memory offsets to reduce stack pressure
+            State memory state;
+            ThreadState memory thread;
+            assembly {
+                state := STATE_MEM_OFFSET
+                thread := TC_MEM_OFFSET
+            }
+
+            // Load the syscall numbers and args from the registers
+            (uint32 syscall_no, uint32 a0, uint32 a1, uint32 a2, uint32 a3) = sys.getSyscallArgs(thread.registers);
+            uint32 v0 = 0;
+            uint32 v1 = 0;
+
+            if (syscall_no == sys.SYS_MMAP) {
+                (v0, v1, state.heap) = sys.handleSysMmap(a0, a1, state.heap);
+            } else if (syscall_no == sys.SYS_BRK) {
+                // brk: Returns a fixed address for the program break at 0x40000000
+                v0 = BRK_START;
+            } else if (syscall_no == sys.SYS_CLONE) {
+                v0 = state.nextThreadID;
+                v1 = 0;
+                ThreadState memory newThread;
+                newThread.threadID = state.nextThreadID;
+                newThread.exitCode = 0;
+                newThread.exited = false;
+                newThread.futexAddr = 0xFF_FF_FF_FF;
+                newThread.futexVal = 0;
+                newThread.futexTimeoutStep = 0;
+                newThread.pc = thread.nextPC;
+                newThread.nextPC = thread.nextPC + 4;
+                newThread.lo = thread.lo;
+                newThread.hi = thread.hi;
+                for (uint256 i; i < 32; i++) {
+                    newThread.registers[i] = thread.registers[i];
+                }
+                newThread.registers[29] = a1; // set stack pointer
+                // the child will perceive a 0 value as returned value instead, and no error
+                newThread.registers[2] = 0;
+                newThread.registers[7] = 0;
+                state.nextThreadID++;
+
+                // Preempt this thread for the new one. But not before updating PCs
+                thread.pc = thread.nextPC;
+                thread.nextPC = thread.nextPC + 4;
```

Comment:
I don't think we're setting the return values for the current thread here, suggest using the helper:
```suggestion
                st.CpuScalars memory cpu0 = getCpuScalars(thread);
                sys.handleSyscallUpdates(cpu0, thread.registers, v0, v1);
```

Reply:
fixed.